src/gp_engine.cpp

```diff
@@ -3,54 +3,171 @@
   quartet_q_s_plv_ = quartet_root_plv_;
   quartet_r_sorted_plv_ = quartet_root_plv_;
+}
 
-  InitializePLVsWithSitePatterns();
+void GPEngine::GrowEngine(const size_t node_count, const size_t gpcsp_count,
+                          std::optional<Reindexer> node_reindexer,
+                          std::optional<Reindexer> gpcsp_reindexer) {
+  GrowPLVs(node_count, node_reindexer);
+  GrowGPCSPs(gpcsp_count, gpcsp_reindexer);
+}
+
+void GPEngine::GrowPLVs(const size_t new_node_count,
+                        std::optional<Reindexer> node_reindexer) {
+  size_t old_node_count = GetNodeCount();
+  size_t old_plv_count = GetPLVCount();
+  node_count_ = new_node_count;
+  // Reallocate more space if needed.
+  if (GetPaddedNodeCount() > GetAllocatedNodeCount()) {
+    node_alloc_ = size_t(ceil(double(GetPaddedNodeCount()) * plv_resizing_factor_));
+    mmapped_master_plv_.Resize(GetAllocatedPLVCount() * site_pattern_.PatternCount());
+    plvs_ = mmapped_master_plv_.Subdivide(GetAllocatedPLVCount());
+    rescaling_counts_.conservativeResize(GetAllocatedPLVCount());
+  }
+  // Initialize new work space.
+  Assert(plvs_.back().rows() == MmappedNucleotidePLV::base_count_ &&
+             plvs_.back().cols() ==
+                 static_cast<Eigen::Index>(site_pattern_.PatternCount()),
+         "Didn't get the right shape of PLVs out of Subdivide.");
+  for (size_t i = old_plv_count; i < GetPaddedPLVCount(); i++) {
+    rescaling_counts_[i] = 0;
+    plvs_.at(i).setZero();
+  }
+  // Reindex work space to match DAG.
+  if (node_reindexer.has_value()) {
+    ReindexPLVs(node_reindexer.value(), old_node_count);
+  }
+}
+
+void GPEngine::GrowGPCSPs(const size_t new_gpcsp_count,
+                          std::optional<Reindexer> gpcsp_reindexer) {
+  size_t old_gpcsp_count = gpcsp_count_;
+  gpcsp_count_ = new_gpcsp_count;
+  // Reallocate more space if needed.
+  if (GetPaddedGPCSPCount() > gpcsp_alloc_) {
+    gpcsp_alloc_ = size_t(ceil(double(GetPaddedGPCSPCount()) * gpcsp_resizing_factor_));
+    branch_lengths_.conservativeResize(gpcsp_alloc_);
+    hybrid_marginal_log_likelihoods_.conservativeResize(gpcsp_alloc_);
+    log_likelihoods_.conservativeResize(gpcsp_alloc_, site_pattern_.PatternCount());
+  }
+  // Initialize new work space.
+  for (size_t i = old_gpcsp_count; i < GetPaddedGPCSPCount(); i++) {
+    branch_lengths_[i] = default_branch_length_;
+    hybrid_marginal_log_likelihoods_[i] = DOUBLE_NEG_INF;
+  }
+  // Reindex work space to match DAG.
+  if (gpcsp_reindexer.has_value()) {
+    ReindexGPCSPs(gpcsp_reindexer.value(), old_gpcsp_count);
+  }
+}
+
+void GPEngine::ReindexPLVs(const Reindexer node_reindexer,
+                           const size_t old_node_count) {
+  Assert(node_reindexer.size() == node_count_,
+         "Node Reindexer is the wrong size for GPEngine.");
+  Assert(node_reindexer.IsValid(node_count_), "Node Reindexer is not valid.");
+
+  // Expand node_reindexer into plv_reindexer.
```

Comment:
I'm not sure why we need this loop based algorithm... I would have thought that we would just expand based on `plv_count_per_node_`.

Reply:
I'm not sure exactly what is meant here?